### ISSUE TITLE
[feature removal] BN_mod_inverse() should reject negative moduli

### ISSUE THREAD

**Issue by romen** [openssl/openssl#16232]
This request was inspired while reviewing #16223.

Technically, for a long time, we have accepted as a valid argument to `BN_mod_inverse()` a modulus `n` even if it happens to be negative. Internally we just flip the `bn->neg` sign flag and continue as if `n` was positive all along.

Similarly it goes for other `BN_mod_*` functions.

There isn't a functional issue with the behavior, but I believe it is desirable to change our behaviour to reject negative moduli:

- this would allow to remove special cases from our implementations and reduce complexity
- I am not aware of real world cryptographic applications where modular arithmetic is done with negative moduli:
  - these special-case codepaths as a result are undertested
  - in most cases, when our users are using `BN_mod_*` directly and the modulus hapens to be negative, most likely something went terribly wrong in the previous steps, but we are potentially masking the problem by handling it
  
I would like to propose for the next major release to explicitly fail with an error code if we detect a negative modulus, rather then flipping the sign and continuing.

**Comment by h-vetinari** [comment]
Should negative moduli be deprecated already now?

**Comment by paulidale** [comment]
Too late for 3.0 IMO.

As a toolkit, we might want to keep the capability.  It should work correctly (in a mathematical sense), which it doesn't seem like it does.

**Comment by romen** [comment]
I agree not for 3.0, because it warrants a wider discussion and we do not have enough time before the final release.
The deprecation would be in the documentation only, without affecting the API, but we really would need time to plan for this properly and discuss how it would affect other BN functions. That is why I am proposing this for next major release.

As a cryptographic toolkit I think it has more potential for harm than for good, and we do not aim to be a general purpose multiprecision arithmetic library.

The current behavior would not be lost anyway, as long as the caller retains the ability to flip the sign of a BIGNUM, it would just need more explicit handling of negative cases.
I see that as a win win win: more security for the cryptographic uses (which do not deal with negative moduli), no loss of functionality for general purpose MPI users (at the cost of a few explicit sign flipping calls), and less complexity inside the library.

**Comment by paulidale** [comment]
Isn't the current behaviour is incorrect for _all_ moduli except -2.

**Comment by romen** [comment]
> Isn't the current behaviour is incorrect for _all_ moduli except -2.

I am not sure I see why it's incorrect.

**Comment by paulidale** [comment]
Wrong sign on my part :(  The modulus sign can be flipped.

Although there are two possible results.  Should the result sign follow the sign of the number or the modulus?  The answer is both are useful in different circumstance -- hence we have both remainder and modulo.

**Comment by romen** [comment]
Yeah, it's the classic option between truncation to 0, a la C, or truncation towards negative infinity, a la Knuth.

Mathematicians usually prefer the latter, as it is easier to express it in terms of mathematical functions using floor. But apart from the notation, both definitions are valid as long as they are applied consistently, and as you mentioned one notation will be more advantageous than the other depending on the application, but going with the C convention as we are doing is the path with least surprises given we are primarily a C library.

I would still like to argue about removal of the feature, for the reasons stated above, as one of the items for whenever a next major release will be in the plans.

Discussing opportunities for simplification is a valuable exercise and a good habit, even if the OTC decided not to remove this feature.

**Comment by paulidale** [comment]
I'm quite partial to round to even :)

**Comment by davidben** [comment]
+1 to rejecting negative moduli. I've been meaning to clear that out on our end, for all the reasons you mention. (We did it for anything using Montgomery reduction without breaking anything, but we still need to go through the others.)

**Comment by romen** [comment]
@nhorman was this considered for the future Major release (which would have been 4.0, given it was moved to that queue in Jun 2024)?

**Comment by nhorman** [comment]
@romen I'll have to apologize here.  I think this was lost in the noise of our issue list, and so this missed for 4.0.  given that we are frozen for beta currently, I don't see this as making it into that release.

On the upside (@mattcaswell @t8m confirm for me please), I think we've been talking about accelerating our major release cycle, and if we are, this will have another chance to get in sooner than our usual 5 year cycle.
